### PR TITLE
Improve `ty.needs_drop`

### PR DESCRIPTION
**Pull request by matthewjasper**
* Handle cycles in `needs_drop` correctly
* Normalize types when computing `needs_drop`
* Move queries from rustc to rustc_ty
* Avoid query in simple cases

reopens #65918

### CODE REVIEW

File: src/librustc/ty/util.rs [hunk 1]
```diff
@@ -974,3 +988,62 @@ impl<'tcx> ExplicitSelf<'tcx> {
...
+        // Pessimistically assume that all generators will require destructors
+        // as we don't know if a destructor is a noop or not until after the MIR
+        // state transformation pass
+        ty::Generator(..) | ty::Dynamic(..) | ty::Error => Err(AlwaysRequiresDrop),
+
+        ty::Slice(ty) => needs_drop_components(ty),
+        ty::Array(elem_ty, ..) => {
+            match needs_drop_components(elem_ty) {
+                Ok(v) if v.is_empty() => Ok(v),
+                // Arrays of size zero don't need drop, even if their element
```
[the remark this answers is not in the document]
Comment (matthewjasper) [reply src/librustc/ty/util.rs]: This was the right arm. The code should be clearer now.

File: src/librustc/ty/util.rs [hunk 2]
```diff
@@ -974,3 +989,72 @@ impl<'tcx> ExplicitSelf<'tcx> {
...
+                Ok(v) if v.is_empty() => Ok(v),
+                res => match size.val.try_to_bits(target_layout.pointer_size) {
+                    // Arrays of size zero don't need drop, even if their element
+                    // type does.
+                    Some(0) => Ok(SmallVec::new()),
+                    Some(_) => res,
+                    // We don't know which of the cases above we are in, so
+                    // return the whole type and let the caller decide what to
+                    // do.
+                    _ => Ok(smallvec![ty]),
```
Comment (bjorn3) [inline src/librustc/ty/util.rs]: s/_/None

**Comment by Zoxc** [comment]
Checking performance impact after https://github.com/rust-lang/rust/pull/68672

**Comment by matthewjasper** [comment]
@rust-timer build d596559b9cb2e0f3efe1e7a4bb8038dfde649f4c

File: src/librustc/ty/mod.rs [hunk 3]
```diff
@@ -2279,6 +2285,12 @@ impl<'tcx> AdtDef {
         self.flags.contains(AdtFlags::IS_BOX)
     }
 
+    /// Returns `true` if this is ManuallyDrop<T>.
```
Comment (varkor) [inline src/librustc/ty/mod.rs]:
```suggestion
    /// Returns `true` if this is `ManuallyDrop<T>`.
```

File: src/librustc/ty/util.rs [hunk 4]
```diff
@@ -974,3 +988,62 @@ impl<'tcx> ExplicitSelf<'tcx> {
...
+        | ty::Never
+        | ty::FnDef(..)
+        | ty::FnPtr(_)
+        | ty::Char
+        | ty::GeneratorWitness(..)
+        | ty::RawPtr(_)
+        | ty::Ref(..)
+        | ty::Str => Ok(SmallVec::new()),
+
+        // Foreign types can never have destructors
```
Comment (varkor) [inline src/librustc/ty/util.rs]:
```suggestion
        // Foreign types can never have destructors.
```

File: src/librustc/ty/util.rs [hunk 5]
```diff
@@ -974,3 +988,62 @@ impl<'tcx> ExplicitSelf<'tcx> {
...
+        ty::Foreign(..) => Ok(SmallVec::new()),
+
+        // Pessimistically assume that all generators will require destructors
+        // as we don't know if a destructor is a noop or not until after the MIR
+        // state transformation pass
```
Comment (varkor) [inline src/librustc/ty/util.rs]:
```suggestion
        // state transformation pass.
```

File: src/librustc_ty/needs_drop.rs [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+use rustc_data_structures::fx::FxHashSet;
+use rustc_hir::def_id::DefId;
+use rustc_span::DUMMY_SP;
+
+type NeedsDropResult<T> = Result<T, AlwaysRequiresDrop>;
+
+fn needs_drop_raw<'tcx>(tcx: TyCtxt<'tcx>, query: ty::ParamEnvAnd<'tcx, Ty<'tcx>>) -> bool {
+    let adt_fields =
+        move |adt_def: &ty::AdtDef| tcx.adt_drop_tys(adt_def.did).map(|tys| tys.iter().copied());
+    // If we don't know a type doesn't need drop, say it's a type parameter
```
Comment (varkor) [inline src/librustc_ty/needs_drop.rs]:
```suggestion
    // If we don't know a type doesn't need drop, for example if it's a type parameter
```

File: src/librustc_ty/needs_drop.rs [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+struct NeedsDropTypes<'tcx, F> {
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    query_ty: Ty<'tcx>,
+    seen_tys: FxHashSet<Ty<'tcx>>,
+    /// A stack of types left to process. Each round, we pop something from the
+    /// stack and check if it needs drop. If the result depends on whether some
+    /// other types need drop we push them onto the stack.
+    unchecked_tys: Vec<(Ty<'tcx>, usize)>,
```
Comment (varkor) [inline src/librustc_ty/needs_drop.rs]: It would be good to document what the `usize` represents here.

**Review by varkor — APPROVED** [review]
Implementation looks good to me. r=me after the comment fixes.

**Comment by matthewjasper** [comment]
Seeing whether caching helps clean incremental or not
@bors try @rust-timer queue

**Comment by jonas-schievink** [comment]
@bors r-

Failed in https://github.com/rust-lang/rust/pull/69056

**Pull request merged**